apps/backend/src/workers/companion-worker.ts

```diff
@@ -0,0 +1,214 @@
+import type { Pool, PoolClient } from "pg"
+import type { PostgresSaver } from "@langchain/langgraph-checkpoint-postgres"
+import { withClient } from "../db"
+import type { CompanionJobData, JobHandler } from "../lib/job-queue"
+import { AuthorTypes, CompanionModes, type AuthorType } from "../lib/constants"
+import type { ProviderRegistry } from "../lib/ai"
+import { StreamRepository } from "../repositories/stream-repository"
+import { MessageRepository } from "../repositories/message-repository"
+import { PersonaRepository, type Persona } from "../repositories/persona-repository"
+import {
+  AgentSessionRepository,
+  SessionStatuses,
+} from "../repositories/agent-session-repository"
+import { runCompanionGraph } from "../agents/companion-runner"
+import { sessionId } from "../lib/id"
+import { logger } from "../lib/logger"
+
+const MAX_CONTEXT_MESSAGES = 20
+
+export interface CompanionWorkerDeps {
+  pool: Pool
+  modelRegistry: ProviderRegistry
+  checkpointer: PostgresSaver
+  serverId: string
+  createMessage: (params: {
+    workspaceId: string
+    streamId: string
+    authorId: string
+    authorType: AuthorType
+    content: string
+  }) => Promise<{ id: string }>
+}
+
+/**
+ * Create the companion job handler for pg-boss.
+ */
+export function createCompanionWorker(
+  deps: CompanionWorkerDeps,
+): JobHandler<CompanionJobData> {
+  const { pool, modelRegistry, checkpointer, serverId, createMessage } = deps
+
+  return async (job) => {
+    const { streamId, messageId, triggeredBy } = job.data
+
+    logger.info(
+      { jobId: job.id, streamId, messageId },
+      "Processing companion job",
+    )
+
+    // Step 1: Load context and create/get session
+    const context = await withClient(pool, async (client) => {
```

Comment:
I feel like we've put too much logic in the wrong place. The worker shouldn't do this, instead I'd like it to take in a dependency on an agent that does most of what we do here.

Think of the worker as an http handler, we wouldn't want to do this much in a handler, would we? The worker should basically pass the job data to an appropriate agent or agent dispatcher or similar which does the persona stuff, session handling, and actual running of things etc. That way, we have a single unit which takes in a minimal context of where it's invoked (e.g., streamId and messageId) that we can reuse and test easier (e.g., in future evals).

Perhaps we can introduce a new `agents` package which is a sort of router that invokes one of the different agents (depending on how we invoked this).